src/persist/src/indexed/cache.rs

```diff
@@ -0,0 +1,156 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! A disk-backed cache for objects in blob storage.
+
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+
+use abomonation::abomonated::Abomonated;
+
+use crate::error::Error;
+use crate::indexed::{BlobFutureBatch, BlobMeta, BlobTraceBatch};
+use crate::storage::Blob;
+
+/// A disk-backed cache for objects in [Blob] storage.
+///
+/// The data for the objects in the cache is stored on disk, mmap'd, and a
+/// validated handle is stored in-memory to avoid repeatedly decoding it.
+///
+/// TODO: Add a limit to bound how much disk this cache can use. The `Arc`
+/// return type for `get_batch` seems correct, but means that a bad user could
+/// starve the cache by indefinitely holding handles. The Arcs could be made
+/// into weak references so the cache could forcefully reclaim the backing data,
+/// but this is going to make performance of using the cached batches
+/// unpredictable. I think we probably want a soft limit and a hard limit where
+/// the soft limit does some alerting and the hard limit starts blocking (or
+/// erroring) until disk space frees up.
+pub struct BlobCache<L: Blob> {
+    blob: Arc<Mutex<L>>,
+    // TODO: Use a disk-backed LRU cache.
+    future: Arc<Mutex<HashMap<String, Arc<BlobFutureBatch>>>>,
+    trace: Arc<Mutex<HashMap<String, Arc<BlobTraceBatch>>>>,
+}
+
+impl<L: Blob> Clone for BlobCache<L> {
```

Comment:
random curiousity: why is this type parameter named `L`? I would have expected `B` maybe?

Reply:
ahhh that makes sense :D